src/environmentd/src/telemetry.rs

```diff
@@ -0,0 +1,160 @@
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Telemetry collection.
+
+// To test this module, you'll need to run environmentd with the following
+// flags.
+//
+//   --tls-mode=disable
+//   --segment-api-key=<REDACTED>
+//   --frontegg-tenant=<REDACTED>
+//   --frontegg-jwk=<REDACTED>
+//   --frontegg-api-token-url=<REDACTED>
+//   --frontegg-password-prefix=mzp_
+//
+// Use values from a personal Materialize Cloud stack for the values that
+// are listed as <REDACTED>.
+//
+// You can then use the Segment debugger to watch the events emitted by your
+// environment in real time:
+// https://app.segment.com/materializeinc/sources/cloud_dev/debugger.
+
+use serde::Serialize;
+use serde_json::json;
+use tokio::time::{self, Duration};
+use tracing::warn;
+use uuid::Uuid;
+
+use mz_ore::collections::CollectionExt;
+use mz_ore::retry::Retry;
+use mz_ore::task;
+
+/// How frequently to send a summary to Segment.
+const REPORT_INTERVAL: Duration = Duration::from_secs(15);
+
+/// Telemetry configuration.
+#[derive(Clone)]
+pub struct Config {
+    /// The Segment client to report telemetry events to.
+    pub segment_client: mz_segment::Client,
+    /// A client to the adapter to introspect.
+    pub adapter_client: mz_adapter::Client,
+    /// The ID of the organization for which to report data.
+    pub organization_id: Uuid,
+}
+
+/// Starts reporting telemetry events to Segment.
+pub fn start_reporting(config: Config) {
+    task::spawn(|| "telemetry_rollup", report_rollup_loop(config.clone()));
+    task::spawn(|| "telemetry_traits", report_traits_loop(config));
+}
+
+async fn report_rollup_loop(
+    Config {
+        segment_client,
+        adapter_client,
+        organization_id,
+    }: Config,
+) {
+    #[derive(Default)]
+    struct Rollup {
+        selects: u64,
+        subscribes: u64,
+    }
+
+    let mut interval = time::interval(REPORT_INTERVAL);
+
+    // The first tick always completes immediately. Ignore it.
+    interval.tick().await;
+
+    let mut last_rollup = Rollup::default();
+    loop {
+        interval.tick().await;
+
+        let query_total = &adapter_client.metrics().query_total;
+        let current_rollup = Rollup {
+            selects: query_total.with_label_values(&["user", "select"]).get(),
+            subscribes: query_total.with_label_values(&["user", "subscribe"]).get(),
+        };
+
+        segment_client.track(
+            // We use the organization ID as the user ID for events
+            // that are not associated with a particular user.
+            organization_id,
+            "Environment Rolled Up",
+            json!({
+                "event_source": "environmentd",
+                "selects": current_rollup.selects - last_rollup.selects,
```

Comment:
I don't know much about segment. From this I assume it doesn't have a native understanding of counters and a rate function?

Reply:
It does not!